### PR TITLE
suggested edits

### PR DESCRIPTION
**Pull request by Kacawi-zz**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,38 +2,37 @@
...
+- You will start with a simple toy example, which can also be found inside a provided docker image.
+- The next step will be to launch a local(!) cluster using `dask.distributed` and run the same grid search again.
+- Lastly, you will learn how to setup a cluster of EC2 *spots* on AWS.
+The last point will be realized in a [declarative](https://en.wikipedia.org/wiki/Declarative_programming) manner using [`terraform`](https://www.terraform.io/).
+
+At the end of this tutorial,  you should have a minimal working example that enable you to run ML processes on a cluster.
+
+## Motivation 
+
+As you read above: we wanted to create a system to analyze and reply to some of them (if applicable). As a first step, a dump of tagged emails was created and the first goal was set: build a Proof Of Concept (POC) that tags the emails automatically.
```
Comment (drorata) [inline README.md]: Maybe stress that "them" = "emails received by our customer service team"?

File: README.md [hunk 2]
```diff
@@ -45,9 +44,10 @@ The project's structure might be:
 └── src
 ```
 
-In `./src` you may include some special tools, functions and classes that you would like to use in the project or in a more complicated pipeline.
-You will see later how to include these tools in the distributed environment; for the time being it is merely a placeholder.
-Note that it is having the structure of a python project and should include a `setup.py` at the project's root.
+In `./src` you may include some special tools, functions and classes that you would like to use in the project or in a more complicated pipeline. You will see later how to include these tools in the distributed environment; for the time being it is merely a placeholder.
```
Comment (drorata) [inline README.md:47]: Side remark. I believe it is a good practice to keep one sentence per line. This makes reviewing and comparing MD files much easier.

File: README.md [hunk 3]
```diff
@@ -119,16 +124,17 @@ RUN pip install -r requirements.txt
...
-It is time to push the image to a Docker registry.
-In this tutorial, you use the AWS service - ECS but you can use other options like `DockerHub`.
-I assume you have [`awscli`](https://aws.amazon.com/cli/) installed, configured and the credentials are known.
-You can log in to the registry simply by
+- It is recommended to include `bokeh` in `requirements.txt`; otherwise the web dashboard of `dask` won't work.
+- The `Dockerfile` can include further steps like `RUN apt-get update && apt-get install -y build-essential freetds-dev` or `RUN python -m nltk.downloader punkt`.
+- If `./src` includes needed classes, functions etc., then make sure you include something like `-e .` or merely `.` in `requirements.txt`; this way these dependencies will be available in the image.
+- It is important to include in the `Dockerfile` all the components needed for the computation environment!
+
+Next, the image should be placed in a location accessible to EC2 instances.It is time to push the image to a Docker registry.
```
Comment (drorata) [inline README.md]: You're missing a 'space'  after the '.' and before 'It'

File: README.md [hunk 4]
```diff
@@ -144,17 +150,19 @@ It is time to setup the nodes of the cluster.
 
 Take a declarative approach and use [`terraform`](https://www.terraform.io) to setup the nodes of the cluster.
 Note that in this example I utilize the AWS Spots; you can easily change the code and use the regular on-demand instances.
-This is left as an exercise.
+
+This is left as an exercise. <!-- maybe give a hint here towards the solution or let people know that the solution is in the GitHub repo --->
```
Comment (drorata) [inline README.md]: Well, the solution is not in the repo... Maybe rephrase to something like:

> Note that in this example I utilize the AWS Spots; it is possible to change the code and use the regular on-demand instances, but it is out of the scope of this tutorial.

In any case, it is an easy exercise :)

File: README.md [hunk 5]
```diff
@@ -144,17 +150,19 @@ It is time to setup the nodes of the cluster.
...
 
 - `.tf` instructions: parsed by `terraform` and defining what instances to use, what tags, regions, etc.
 - Provisioning shell scripts: installing needed tools on the nodes
 
 ### `.tf` files
 
-When using `terraform` all `.tf` files are read and concatenated.
-There are more details of course; a good entry point would be [here](https://www.terraform.io/docs/configuration/index.html).
-In our example I organized the `.tf` files as follows:
+When using `terraform` all `.tf` files are read and concatenated.There are more details of course; a good entry point would be [here](https://www.terraform.io/docs/configuration/index.html).
```
Comment (drorata) [inline README.md]: Again, missing a '.'

Comment (drorata) [inline README.md]: Again, missing a '.'

File: README.md [hunk 6]
```diff
@@ -385,20 +390,15 @@ terraform apply -var 'workersNum=2' -var 'instanceType="t2.small"' \
...
 ```
 
-Note that use of two environment variables for the AWS keys.
-Other variables defined in `var.tf` are passed as parameters.
-Once finished, you can access the newly created scheduler node by: `ssh -i ~/.aws/key.pem ec2-user@$(terraform output scheduler-info)`.
-For this to work, you might need to download a `pem` file from AWS console.
-In the cluster you can check the log at `/var/log/user-data.log`.
-You can also check the status of the running Docker containers using `docker ps`.
-Lastly, if everything went well, you should be able to access the web interface of the cluster.
+**Note** that use of two environment variables for the AWS keys. Other variables defined in `var.tf` are passed as parameters.
```
Comment (drorata) [inline README.md]: Should be:

> **Note** the use of two environment variables....

File: README.md [hunk 7]
```diff
@@ -385,20 +390,15 @@ terraform apply -var 'workersNum=2' -var 'instanceType="t2.small"' \
...
+
+<!-- could you restructure this paragraph a bit? --->
```
Comment (drorata) [inline README.md]: Let's have another iteration on this paragraph after we merge your changes. OK?

File: README.md [hunk 8]
```diff
@@ -453,14 +451,15 @@ docker run -it --rm -p 8786:8786 dask-example ./gridsearch_cluster_dask.py x.y.z
...
 
 I hope that this tutorial gave you an idea how to start a cluster on AWS for distributed computing using `dask`.
-This is obvisouly a mere starting point.
-In many real problems, you would have to also play smart and you won't get away by simply spinning more and more nodes.
+This is obvisouly a mere starting point. In many real problems, you would have to also play smart and you won't get away by simply spinning more and more nodes.
+
 A side story in this tutorial is the beauty of a declarative approach.
-If you make sure that your environment is well defined using `environment.yml` or `requirements.txt` together with proper `Dockerfile` you are on a winning track.
-Reproducing your work is going to be much easier and scaling should become more straightforward as well.
+
```
Comment (drorata) [inline README.md]: Is the new paragraph needed?

**Review by drorata — COMMENTED** [review]
@Kacawi please have a look at the comments I left. Let's conclude this PR (except the paragraph that you suggested to edit which will tackle afterward)

**Comment by drorata** [comment]
@Kacawi reminder that I have reviewed your suggestions.

File: README.md (hunk 1 above)
Comment (Kacawi-zz) [reply README.md]: 👍

File: README.md (hunk 2 above)
Comment (Kacawi-zz) [reply README.md:47]: Sorry; I already adjusted this because it would look horrible when published on our platform.

File: README.md (hunk 8 above)
Comment (Kacawi-zz) [reply README.md]: Not necessarily; Added this for a better reading flow on our blog.

File: README.md (hunk 5 above)
Comment (Kacawi-zz) [reply README.md]: Not sure what you mean by this?

**Review by Kacawi-zz — COMMENTED** [review]
@drorata I have reviewed your comments.

File: README.md (hunk 2 above)
Comment (drorata) [reply README.md:47]: I am surprised. Line breaks have no effect when using markdown...

File: README.md [hunk 9]
```diff
@@ -14,7 +14,7 @@ At the end of this tutorial,  you should have a minimal working example that ena
 
 ## Motivation 
 
-As you read above: we wanted to create a system to analyze and reply to some of them (if applicable). As a first step, a dump of tagged emails was created and the first goal was set: build a Proof Of Concept (POC) that tags the emails automatically. 
+As you read above: we wanted to create a system to analyze and reply to some of emails received by our customer service team (if applicable). As a first step, a dump of tagged emails was created and the first goal was set: build a Proof Of Concept (POC) that tags the emails automatically.
```
Comment (drorata) [inline README.md]: Should be "...reply to some of the emails..."

**Comment by drorata** [comment]
@Kacawi I guess we're almost done.

**Comment by drorata** [comment]
@Kacawi By "almost done" I meant that I left one more comment on this PR. Once resolved I'll merge it. Then, we're left with one paragraph to fix/change if I remember correctly.

**Comment by Kacawi-zz** [comment]
@drorata If you referred to the last comment on the new paragraph, I have removed this one.

File: README.md (hunk 9 above)
Comment (drorata) [reply README.md]: @Kacawi can you fix this typo? I believe it is one

File: README.md (hunk 7 above)
Comment (drorata) [reply README.md]: @Kacawi I meant this paragraph that you still want to edit. I will create a PR once we're done with this one.

File: README.md (hunk 5 above)
Comment (drorata) [reply README.md]: @Kacawi I meant a missing space between "concatenated." and "There". But it seems like you fixed it.

**Comment by drorata** [comment]
@Kacawi I updated the paragraph we discussed. Please have a look at #2

**Comment by drorata** [comment]
@Kacawi I believe #2 is the step, right? Anything else?
